File: Word.cpp
//
// Created by alexs on 3/29/2025.
//

#include "Word.h"
#include "Logger.h"

#include <cstring>
#include <iomanip>
#include <ios>
#include <iosfwd>
#include <iostream>

Word::Word(const int n) {
    saveInHex(n);
}

Word::Word(const std::string &str) {
    memcpy(word, str.c_str(), WORD_SIZE);
}

void Word::saveInHex(const int& n) {
    char buffer[WORD_SIZE + 1];
    snprintf(buffer, sizeof(buffer), "%06X", n);
    memcpy(word, buffer, WORD_SIZE);
}

bool Word::isNumber() const {
    for (int i = 0; i < WORD_SIZE; i++) {
        if (!(word[i] >= '0' && word[i] <= '9') && !(word[i] >= 'A' && word[i] <= 'F')) {
            return false;
        }
    }

    return true;
}

int Word::toInteger() const {
    if (!isNumber()) {
        Logger::debug("Word::toInteger: Invalid number: %.6s", word);
    }

    std::stringstream ss;
    ss << std::hex << std::string(word, WORD_SIZE);
    int n;
    ss >> n;
    return n;
}

bool Word::equals(std::string str) const {
    return memcmp(word, str.c_str(), WORD_SIZE) == 0;
}

bool Word::startsWith(std::string str) const {
    return memcmp(word, str.c_str(), str.size()) == 0;
}

std::string Word::substring(int index) const {
    if (index < 0 || index >= WORD_SIZE) {
        Logger::debug("Word::substring: Invalid index: %d", index);
    }

    return std::string(word).substr(index);
}

void Word::operator--() {
    if (!isNumber()) {
        Logger::debug("Word::operator--: Invalid number: %.6s", word);
    }

    saveInHex(toInteger() - 1);
}

void Word::operator++() {
    if (!isNumber()) {
        Logger::debug("Word::operator++: Invalid number: %.6s", word);
    }

    saveInHex(toInteger() + 1);
}

Word Word::operator+(const Word& other) const {
    if (!isNumber() || !other.isNumber()) {
        Logger::debug("Word::operator+: Invalid number: %.6s", word);
    }

    return Word((toInteger() + other.toInteger()) % 0x1000000);
}

Word Word::operator-(const Word& other) const {
    if (!isNumber() || !other.isNumber()) {
        Logger::debug("Word::operator+: Invalid number: %.6s", word);
    }

    int i1 = toInteger();
    int i2 = other.toInteger();

    if (i1 < i2) {
        i1 += 0x1000000;
    }

    return Word(i1 - i2);
}

Word Word::operator*(const Word& other) const {
    if (!isNumber() || !other.isNumber()) {
        Logger::debug("Word::operator+: Invalid number: %.6s", word);
    }

    return Word(toInteger() * other.toInteger());
}

Word Word::operator/(const Word& other) const {
    if (!isNumber() || !other.isNumber()) {
        Logger::debug("Word::operator+: Invalid number: %.6s", word);
    }

    return Word(toInteger() / other.toInteger());
}

Word Word::operator&(const Word& other) const {
    if (!isNumber() || !other.isNumber()) {
        Logger::debug("Word::operator+: Invalid number: %.6s", word);
    }

    return Word(toInteger() & other.toInteger());
}

Word Word::operator|(const Word& other) const {
    if (!isNumber() || !other.isNumber()) {
        Logger::debug("Word::operator+: Invalid number: %.6s", word);
    }

    return Word(toInteger() | other.toInteger());
}

Word Word::operator^(const Word& other) const {
    if (!isNumber() || !other.isNumber()) {
        Logger::debug("Word::operator+: Invalid number: %.6s", word);
    }

    return Word(toInteger() ^ other.toInteger());
}

std::ostream& operator<<(std::ostream& os, const Word& word) {
    for (int i = 0; i < 6; ++i)
        os << word.word[i];
    return os;
}

File: src/Word.java
import utils.Logger;

import java.util.Arrays;

public class Word {
    private final char[] word = new char[Constants.WORD_SIZE];

    public Word() {
        Arrays.fill(word, '0');
    }

    public Word(int n) {
        saveInHex(n);
    }

    public Word(String str) {
        if (str.length() > Constants.WORD_SIZE) {
            Arrays.fill(word, '0');
        } else {
            for (int i = 0; i < Constants.WORD_SIZE; i++) {
                word[i] = str.charAt(i);
            }
        }
    }

    private void saveInHex(int n) {
        String hex = String.format("%06X", n & 0xFFFFFF);  // limit to 24 bits
        for (int i = 0; i < Constants.WORD_SIZE; i++) {
            word[i] = hex.charAt(i);
        }
    }

    public int toInteger() {
        return Integer.parseInt(new String(word), 16);
    }

    public boolean equals(String str) {
        return new String(word).equals(str);
    }

    public boolean startsWith(String str) {
        return new String(word).startsWith(str);
    }

    public String substring(int index) {
        if (index < 0 || index >= Constants.WORD_SIZE) {
            Logger.debug("Word.substring: Invalid index: " + index);
        }
        return new String(word).substring(index);
    }

    public void decrement() {
        saveInHex(toInteger() - 1);
    }

    public void increment() {
        saveInHex(toInteger() + 1);
    }

    public Word add(Word other) {
        return new Word((this.toInteger() + other.toInteger()) % 0x1000000);
    }

    public Word subtract(Word other) {
        int i1 = this.toInteger();
        int i2 = other.toInteger();
        if (i1 < i2) {
            i1 += 0x1000000;
        }
        return new Word(i1 - i2);
    }

    public Word multiply(Word other) {
        return new Word(this.toInteger() * other.toInteger());
    }

    public Word divide(Word other) {
        return new Word(this.toInteger() / other.toInteger());
    }

    public Word and(Word other) {
        return new Word(this.toInteger() & other.toInteger());
    }

    public Word or(Word other) {
        return new Word(this.toInteger() | other.toInteger());
    }

    public Word xor(Word other) {
        return new Word(this.toInteger() ^ other.toInteger());
    }

    @Override
    public String toString() {
        return new String(word);
    }
}
